### PR TITLE
Adds a validity check for configclasses

### PR DESCRIPTION
**Pull request by Dhoeller19**
# Description

Added a mechanism to check for the validity of a configclass object.
A configclass object is valid if it contains no MISSING attributes.

## Type of change

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [x] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

### CODE REVIEW

**Review by lgulich — APPROVED** [review]
Lgtm, ooi did you ever consider to use pydantic for this? It could probably do many things out of the box but may not be flexible enough for some very advanced use cases.

File: source/extensions/omni.isaac.lab/omni/isaac/lab/utils/configclass.py [hunk 1]
```diff
@@ -98,6 +98,7 @@ class EnvCfg:
     setattr(cls, "from_dict", _update_class_from_dict)
     setattr(cls, "replace", _replace_class_with_kwargs)
     setattr(cls, "copy", _copy_class)
+    setattr(cls, "assert_valid", _assert_valid)
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/utils/configclass.py]: NIT: Maybe a simpler name?

```suggestion
    setattr(cls, "validate", _validate)
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/utils/configclass.py [hunk 2]
```diff
@@ -240,6 +241,55 @@ class State:
     cls.__annotations__ = hints
 
 
+def _assert_valid(obj: object, prefix: str = "") -> list:
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/utils/configclass.py]: NIT: Might be nicer to move this function lower to keep all "init" level things together

```suggestion
def _validate(obj: object, prefix: str = "") -> list[str]:
```

File: source/extensions/omni.isaac.lab/test/utils/test_configclass.py [hunk 3]
```diff
@@ -888,6 +941,18 @@ def test_config_md5_hash(self):
...
 
+    def test_validity(self):
+        """Check that invalid configurations raise errors."""
+
+        cfg = MissingChildDemoCfg()
+
+        with self.assertRaises(TypeError) as context:
+            cfg.assert_valid()
+
+        error_message = str(context.exception)
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/test/utils/test_configclass.py:952]:
```suggestion
            cfg.assert_valid()

            error_message = str(context.exception)
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/utils/configclass.py [hunk 4]
```diff
@@ -7,14 +7,14 @@
...
-from collections.abc import Callable
+from collections.abc import Callable, Mapping
 from copy import deepcopy
 from dataclasses import MISSING, Field, dataclass, field, replace
 from typing import Any, ClassVar
 
 from .dict import class_to_dict, update_class_from_dict
 
-_CONFIGCLASS_METHODS = ["to_dict", "from_dict", "replace", "copy"]
+_CONFIGCLASS_METHODS = ["to_dict", "from_dict", "replace", "copy", "assert_valid"]
```
Comment (Dhoeller19) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/utils/configclass.py]:
```suggestion
_CONFIGCLASS_METHODS = ["to_dict", "from_dict", "replace", "copy", "validate"]
```

File: source/extensions/omni.isaac.lab/test/utils/test_configclass.py [hunk 5]
```diff
@@ -888,6 +941,18 @@ def test_config_md5_hash(self):
 
         self.assertEqual(md5_hash_1, md5_hash_2)
 
+    def test_validity(self):
+        """Check that invalid configurations raise errors."""
+
+        cfg = MissingChildDemoCfg()
+
+        with self.assertRaises(TypeError) as context:
+            cfg.assert_valid()
```
Comment (Dhoeller19) [inline source/extensions/omni.isaac.lab/test/utils/test_configclass.py]:
```suggestion
            cfg.validate()
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/terrains/terrain_importer_cfg.py [hunk 6]
```diff
@@ -36,7 +36,7 @@ class TerrainImporterCfg:
     All sub-terrains are imported relative to this prim path.
     """
 
-    num_envs: int = MISSING
+    num_envs: int = 1
     """The number of environment origins to consider."""
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/terrains/terrain_importer_cfg.py]:
```suggestion
    """The number of environment origins to consider. Defaults to 1.
    
    In case the :class:`~omni.isaac.lab.scene.InteractiveSceneCfg` is used, this parameter gets overidden by
    :class:`~omni.isaac.lab.scene.InteractiveSceneCfg.num_envs` attribute.
    """
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/terrains/terrain_generator_cfg.py [hunk 7]
```diff
@@ -88,7 +88,7 @@ class SubTerrainBaseCfg:
     is 0.7.
     """
 
-    size: tuple[float, float] = MISSING
+    size: tuple[float, float] = (10.0, 10.0)
     """The width (along x) and length (along y) of the terrain (in m)."""
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/terrains/terrain_generator_cfg.py]:
```suggestion
    """The width (along x) and length (along y) of the terrain (in m). Defaults to (10.0, 10.0).
    
    In case the :class:`~omni.isaac.lab.terrains.TerrainImporterCfg` is used, this parameter gets overidden by
    :attr:`omni.isaac.lab.scene.TerrainImporterCfg.size` attribute.
    """
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/terrains/terrain_importer_cfg.py [hunk 8]
```diff
@@ -36,8 +36,12 @@ class TerrainImporterCfg:
     All sub-terrains are imported relative to this prim path.
     """
 
-    num_envs: int = MISSING
-    """The number of environment origins to consider."""
+    num_envs: int = 1
+    """The number of environment origins to consider. Defaults to 1.
+    
+    In case, the :class:`~omni.isaac.lab.scene.InteractiveSceneCfg` is used, this parameter gets overidden by
+    :class:`~omni.isaac.lab.scene.InteractiveSceneCfg.num_envs` attribute.
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/terrains/terrain_importer_cfg.py]:
```suggestion
    :attr:`omni.isaac.lab.scene.InteractiveSceneCfg.num_envs` attribute.
```

File: docs/source/tutorials/03_envs/create_manager_rl_env.rst [hunk 9]
```diff
@@ -117,11 +117,10 @@ For various goal-conditioned tasks, it is useful to specify the goals or command
...
 commands at each step. It can also be used to provide the commands as an observation to the agent.
 
-For this simple task, we do not use any commands. This is specified by using a command term with the
-:class:`envs.mdp.NullCommandCfg` configuration. However, you can see an example of command definitions in the
-locomotion or manipulation tasks.
+In this tutorial we don't implement commands for simplicity, but you can see an example of commands definition in the
+other locomotion or manipulation tasks.
 
-.. literalinclude:: ../../../../source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/manager_based/classic/cartpole/cartpole_env_cfg.py
+.. literalinclude:: ../../../../source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/manager_based/locomotion/velocity/velocity_env_cfg.py
```
Comment (Mayankm96) [inline docs/source/tutorials/03_envs/create_manager_rl_env.rst]: I think it is confusing to refer to this snippet when we're talking about cartpole environment. Shall we omit it out completely?

File: source/extensions/omni.isaac.lab/docs/CHANGELOG.rst [hunk 10]
```diff
@@ -1,6 +1,25 @@
 Changelog
 ---------
 
+0.26.1 (2024-10-14)
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/docs/CHANGELOG.rst]: Major bump since we break stuff....

Comment (Mayankm96) [reply source/extensions/omni.isaac.lab/docs/CHANGELOG.rst]: Can we make it somewhat backwards compatible? So throw a warning if someone uses NullCommand class, and we tell them to set command to None instead? Ideally if you are just playing around, you may just set it to NullCommand for testing and then change it to something else once you are ready.

Changing the variable to None seems more effortful.

File: source/extensions/omni.isaac.lab/omni/isaac/lab/assets/asset_base_cfg.py [hunk 11]
```diff
@@ -39,8 +39,8 @@ class InitialStateCfg:
         Defaults to (1.0, 0.0, 0.0, 0.0).
         """
 
-    class_type: type[AssetBase] = MISSING
-    """The associated asset class.
+    class_type: type[AssetBase] = None
+    """The associated asset class. Defaults to None.
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/assets/asset_base_cfg.py]: What does None mean? Would be better to clarify.

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/commands/velocity_command.py [hunk 12]
```diff
@@ -67,6 +69,19 @@ def __init__(self, cfg: UniformVelocityCommandCfg, env: ManagerBasedEnv):
         self.metrics["error_vel_xy"] = torch.zeros(self.num_envs, device=self.device)
         self.metrics["error_vel_yaw"] = torch.zeros(self.num_envs, device=self.device)
 
+        if self.cfg.heading_command:
+            if self.cfg.ranges.heading is None:
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/commands/velocity_command.py]:
```suggestion
        if self.cfg.heading_command and self.cfg.ranges.heading is None:
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/commands/velocity_command.py [hunk 13]
```diff
@@ -67,6 +69,19 @@ def __init__(self, cfg: UniformVelocityCommandCfg, env: ManagerBasedEnv):
...
+                raise ValueError(
+                    "The velocity command has heading commands active (heading_command=True) but the `ranges.heading`"
+                    " parameter is set to None."
+                )
+        if self.cfg.ranges.heading and not self.cfg.heading_command:
+            omni.log.warn(
+                "The velocity command configuration has the ranges.heading attribute set to"
+                f" {self.cfg.ranges.heading} but the heading command is not active. Consider setting"
+                " heading_command to True."
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/commands/velocity_command.py]:
```suggestion
                " the flag for the heading command to True."
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/commands/velocity_command.py [hunk 14]
```diff
@@ -67,6 +69,19 @@ def __init__(self, cfg: UniformVelocityCommandCfg, env: ManagerBasedEnv):
...
 
+        if self.cfg.heading_command:
+            if self.cfg.ranges.heading is None:
+                raise ValueError(
+                    "The velocity command has heading commands active (heading_command=True) but the `ranges.heading`"
+                    " parameter is set to None."
+                )
+        if self.cfg.ranges.heading and not self.cfg.heading_command:
+            omni.log.warn(
+                "The velocity command configuration has the ranges.heading attribute set to"
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/commands/velocity_command.py]:
```suggestion
                "The velocity command configuration has the 'ranges.heading' attribute set to"
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/sensors/ray_caster/ray_caster.py [hunk 15]
```diff
@@ -48,14 +48,6 @@ class RayCaster(SensorBase):
 
     cfg: RayCasterCfg
     """The configuration parameters."""
-    meshes: ClassVar[dict[str, wp.Mesh]] = {}
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/sensors/ray_caster/ray_caster.py:51]: Why was this moved from class variable to instance variable? I thought the reasoning was to not recreate warp meshes if they exist in memory?

Comment (Mayankm96) [reply source/extensions/omni.isaac.lab/omni/isaac/lab/sensors/ray_caster/ray_caster.py:51]: @pascal-roth for vis.

**Review by Mayankm96 — COMMENTED** [review]
Looks fine. A huge MR but the changes just had to be rolled out everywhere so it's okay (not ideal though).

only concern left is ray-caster change. Would be great to isolate these changes because they can't be "clubbed" into other changes...

File: source/extensions/omni.isaac.lab/docs/CHANGELOG.rst [hunk 16]
```diff
@@ -1,6 +1,21 @@
...
+^^^^^
+
+* Added a method to :class:`~omni.isaac.lab.utils.configclass` to check for attributes with values of
+  type ``MISSING``. This is useful when the user wants to check if a certain attribute has been set or not.
+* Added the configuration validation check inside the constructor of all the core classes
+  (such as sensor base, asset base, scene and environment base classes).
+* Added support for environments without commands by leaving the attribute
+  :attr:`omni.isaac.lab.envs.ManagerBasedRLEnvCfg.commands` as None. Before, this had to be done using
+  the class :class:`omni.isaac.lab.command_generators.NullCommandGenerator`.
+
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/docs/CHANGELOG.rst]: Another change that needs attention if we go with it

```suggestion
  the class :class:`omni.isaac.lab.command_generators.NullCommandGenerator`.
* Moved the ``meshes`` attribute in the :class:`omni.isaac.lab.sensors.RayCaster` class from class variable to instance variable.
   This prevents the meshes to overwrite each other.

```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/sensors/ray_caster/ray_caster.py (hunk 15 above)
Comment (Dhoeller19) [reply source/extensions/omni.isaac.lab/omni/isaac/lab/sensors/ray_caster/ray_caster.py:51]: The reason is they do not get deallocated when the env is closed. So when you loop through environments (for example in a grid search) it might never create a new mesh.

**Review by Mayankm96 — APPROVED** [review]
